articles/150_roslaunch.md

```diff
@@ -0,0 +1,364 @@
+
+Alternatively, we could say "all event handling is opaque to the launch system and is user defined" which removes the ambiguity.
+
+\<wjwwod's opinion>
+<br/>
+I think the idea that saying all "supervision" or "reactionary launching" like activities are not part of the launch system is the simplest way forward, but then you'd still like to include in the launch system some convenient ways to express the most common cases, e.g. something like ROS 1's `roslaunch`'s `require=true` and `respawn=true`.
+In my opinion, it would be unacceptably inconvenient to have users write a lambda for each of those cases, and I also think it would lead to everyone doing it slightly different and making it impossible to understand with tools.
+<br/>
+</wjwwod's opinion>
+
+This will come up again in the "System Description" section, because I believe the mind set around the "Execution of the System Description" (basically the python library that runs things) will be "personal responsibility".
+</div>
+
+#### Termination
+
+Managed ROS Nodes have some additional observable effects when terminating (the node, not necessarily the process containing it).
+A managed node enters the `Finalized` state after passing through the `ShuttingDown` transition state on termination.
+Since these are state transitions, they are observable via the lifecylce event system, at least through the ROS topic `lifecycle_state` (subject to change, always reference the managed nodes design document[^lifecylce]).
+
+### Process with a Single Node
+
+In this subsection, and the following subsections of the "Calling Conventions" section, the different possible combinations of nodes and processes is explained.
+In each case they "inherit" any behaviors from either the "ROS nodes" or the "Managed ROS nodes" subsections above, but in these subsections the "how" of communicating ROS specific options is described in more detail.
+
+The first is a single process with a single ROS node within it.
+This was likely the most commonly used type of entity launched in ROS 1, as you could only have one node per process in ROS 1.
+In ROS 2, this will likely be less common because you can have one to many nodes per process, but will may still be used quite a bit in the form of quickly developed scripts and drivers or GUI tools which might require control over the main thread.
+
+Since there is only one ROS node, the command line arguments do not need to be explicit about to which node they apply.
+For example, changing the namespace of the single node could be expressed with the command line argument `__ns:=new_namespace`.
+
+Even though there is only one node in the process, that node does not need to start with the process starts, nor does the process need to end when the node is shutdown and/or destroyed.
+If it is a managed node, the lifecycle of the node is best tracked using the lifecycle events.
```

Comment:
Is ROS 2 launch responsible for transitioning a node to `Active`? If so, how would it know if a process contained a normal node or a managed one? If not, how would the user manage the node? By launching a separate node that triggers the transitions?

Reply:
> Is ROS 2 launch responsible for transitioning a node to Active?

Again, not sure. My initial reaction is "no, the node should transition itself unless directed to do otherwise". But I haven't fleshed out how that would work. We (me, Karsten, perhaps others) have spoken in the past about lifecycle profiles which would be basically an option to lifecycle nodes which determines what they do automatically and what they wait to do based on external requests, e.g. a default profile might transition all the way to active, then stay there until shutdown (sigint or something else), or a managed profile that goes automatically until inactive then waits for external transition requests or shutdown.

> If so, how would it know if a process contained a normal node or a managed one?

I believe the user will already have to inform the launch system if a process contains a node and if so whether or not it is a managed node. So, the user would have to tell it, otherwise it will assume it's a normal node or a normal process.

> If not, how would the user manage the node? By launching a separate node that triggers the transitions?

Yes, this gets into one of the trickier issues for which I've been trying to come up with the right taxonomy (meronomy?) relationships. Basically, where does the launch system end and general purpose lifecycle managed begin? The launch system could use the lifecycle to verify the system and/or do more sophisticated startup procedures (e.g. when nodes A and B get to state inactive, launch node C...), but how is that different from (if node A goes to error state, move node B to inactive). Should it do both, or neither, or some grey line in between?

If the answer is neither or some grey line in between, then yes the user would simply start another node to manage the nodes that need external transitions.

Reply:
It sort of sounds like you're saying your gut reaction is to separate the concern of launching nodes from that of managing their lifecycle.

Assuming a separate process/project for managing lifecycle nodes, it seems like that project would need an API into ROS 2 launch rather than the other way around. E.g. a managing process might ask ROS 2 launch to start some unmanaged log-backup node when a managed node transitions to `cleanup`.

Reply:
> Assuming a separate process/project for managing lifecycle nodes, it seems like that project would need an API into ROS 2 launch rather than the other way around. E.g. a managing process might ask ROS 2 launch to start some unmanaged log-backup node when a managed node transitions to cleanup.

Yeah, I expect we'll have that interface regardless. It should be possible to interact with a running launch system via API (probably Python), or via ROS interfaces like services, or maybe even via a different SOA like REST or XMLRPC (just ideas, no concrete plans to do that).

Reply:
I think there's already a note that lifecycle stuff is out of scope for this document and how it is actually achieved and by who (launch, user, autonomously?) will be determined by some other lifecycle specific documentation.